cns/service.go

```diff
@@ -48,63 +49,79 @@ func NewService(name, version, channelMode string, store store.KeyValueStore) (*
 	}, nil
 }
 
-// GetAPIServerURL returns the API server URL.
-func (service *Service) getAPIServerURL() string {
-	urls, _ := service.GetOption(acn.OptCnsURL).(string)
-	if urls == "" {
-		urls = defaultAPIServerURL
+// AddListeners is to add two listeners(nodeListener and localListener) for connections on the given address
+func (service *Service) AddListeners(config *common.ServiceConfig, primaryIP string) error {
+	var nodeAPIServerURL *url.URL
+
+	// Fetch and parse the API server URL.
+	if service.GetOption(acn.OptCnsURL).(string) == "" {
+		// get VM primary interface's private IP
+		nodeAPIServerURL, _ = url.Parse(fmt.Sprintf("tcp://%s:%s", primaryIP, defaultAPIServerPort))
+	} else {
+		nodeAPIServerURL, _ = url.Parse(service.GetOption(acn.OptCnsURL).(string))
+	}
+
+	// construct url
+	nodeListener, err := acn.NewListener(nodeAPIServerURL)
+	if err != nil {
+		return errors.Wrap(err, "Failed to construct url for node listener")
 	}
 
-	return urls
+	// only use TLS connection for DNC/CNS listener:
+	if config.TlsSettings.TLSPort != "" {
+		// listener.URL.Host will always be hostname:port, passed in to CNS via CNS command
+		// else it will default to localhost
+		// extract hostname and override tls port.
+		hostParts := strings.Split(nodeListener.URL.Host, ":")
+		tlsAddress := net.JoinHostPort(hostParts[0], config.TlsSettings.TLSPort)
+
+		// Start the listener and HTTP and HTTPS server.
+		tlsConfig, err := getTLSConfig(config.TlsSettings, config.ErrChan) //nolint
+		if err != nil {
+			log.Printf("Failed to compose Tls Configuration with error: %+v", err)
+			return errors.Wrap(err, "could not get tls config")
+		}
+
+		if err := nodeListener.StartTLS(config.ErrChan, tlsConfig, tlsAddress); err != nil {
+			return errors.Wrap(err, "could not start tls")
+		}
+	}
+
+	nodeListener.ListenerType = "nodeListener"
+	*config.Listeners = append(*config.Listeners, *nodeListener)
```

Comment:
once you make the above change, this should just be `config.Listeners = append(config.Listeners, nodeListener)`, since `acn.NewListener` returned a listener pointer already